sklearn/manifold/tests/test_isomap.py

```diff
@@ -116,6 +116,33 @@ def test_pipeline():
     assert_less(.9, clf.score(X, y))
 
 
+def test_pipeline_with_nearest_neighbors_transformer():
+    # Test chaining NearestNeighborsTransformer and Isomap with
+    # neighbors_algorithm='precomputed'
+    algorithm = 'auto'
+    n_neighbors = 10
+
+    X, _ = datasets.make_blobs(random_state=0)
+    X2, _ = datasets.make_blobs(random_state=1)
+
+    # compare the chained version and the compact version
+    est_chain = pipeline.make_pipeline(
+        neighbors.KNeighborsTransformer(
+            n_neighbors=n_neighbors + 1, algorithm=algorithm, mode='distance'),
+        manifold.Isomap(n_neighbors=n_neighbors,
+                        neighbors_algorithm='precomputed'))
+    est_compact = manifold.Isomap(n_neighbors=n_neighbors,
```

Comment:
To avoid the issue that one needs to know that n neighbors is +1 for isomap can we ask that n_neigbors is None if a sparse graph is passed ?

Reply:
We can.
But one benefit of this PR is that you can cache the neighbors graph computation, and try multiple `n_neighbors` for instance in a `GridSearchCV` without recomputing the graph. If the number of neighbors in the precomputed graph is larger than necessary, we filter to keep only the closest neighbors (cf [here](https://github.com/scikit-learn/scikit-learn/pull/10482/files#diff-3d65710180ff5cb7d9ffb176e1ad721aR228)).
If we use `n_neighbors=None`, we remove this possibility and ask the subsequent estimators to use all precomputed neighbors.

The graph filtering makes having extra neighbors painless, which is why I proposed to compute one extra neighbor in `KNeighborsTransformer`, always or only when `self.mode='distance'`. This is similar to the magic behind `kneighbors(X=None)`.